packages/react-dom/src/__tests__/ReactMount-test.js

```diff
@@ -160,6 +169,306 @@ describe('ReactMount', () => {
+
+    class Component extends React.Component {
+      render() {
+        return this.props.children;
+      }
+    }
+
+    const serverRandom = Math.random();
+    const clientRandom = Math.random();
+
+    const div = document.createElement('div');
+    const markup = ReactDOMServer.renderToString(
+      <Component>
+        <em>
+          {'SSRMismatchTest static text and '}
+          {'server random text ' + serverRandom}
+        </em>
+      </Component>,
+    );
+    div.innerHTML = markup;
+
+    expect(() =>
+      ReactDOM.hydrate(
+        <Component>
+          <em>
+            {'SSRMismatchTest static text and '}
+            {'client random text ' + clientRandom}
+          </em>
+        </Component>,
+        div,
+      ),
+    ).toWarnDev(
+      'Text content did not match. ' +
+        'Server: "server random text ' +
+        serverRandom +
+        '" ' +
+        'Client: "client random text ' +
+        clientRandom +
+        '"\n' +
+        '    in em (at **)\n' +
+        '    in Component (at **)',
+    );
+  });
+
+  it('should warn when a hydrated element has children mismatch (replacement diff)', () => {
+    // See fixtures/ssr: ssr-warnForInsertedHydratedTextInstance-didNotFindHydratableTextInstance-replacement
+
+    class Component extends React.Component {
+      render() {
+        return this.props.children;
+      }
+    }
+
+    const div = document.createElement('div');
+    const markup = ReactDOMServer.renderToString(
+      <Component>
+        nested{'   '}
+        <h1>
+          children <b>text</b>
+        </h1>
+        <div data-ssr-mismatch-padding-after="1" />
+        <div data-ssr-mismatch-padding-after="2" />
+        <div data-ssr-mismatch-padding-after="3" />
+        <div data-ssr-mismatch-padding-after="4" />
+        <div data-ssr-mismatch-padding-after="5" />
+      </Component>,
+    );
+    div.innerHTML = markup;
+
+    expect(div.outerHTML).toEqual(
+      '<div>nested<!-- -->   <h1>children <b>text</b></h1>' +
+        '<div data-ssr-mismatch-padding-after="1"></div>' +
+        '<div data-ssr-mismatch-padding-after="2"></div>' +
+        '<div data-ssr-mismatch-padding-after="3"></div>' +
+        '<div data-ssr-mismatch-padding-after="4"></div>' +
+        '<div data-ssr-mismatch-padding-after="5"></div>' +
+        '</div>',
+    );
+
+    expect(() =>
+      ReactDOM.hydrate(
+        <Component>
+          nested{'   '}
+          <h2>
+            children <b>text</b>
+          </h2>
+        </Component>,
+        div,
+      ),
+    ).toWarnDev(
+      'Warning: Expected server HTML to contain a matching <h2> in <div>.\n\n' +
+        '  <div>\n' +
+        "    {'nested'}\n" +
+        '    <!-- -->\n' +
+        "    {'   '}\n" +
+        '-   <h1>children <b>text</b></h1>\n' +
+        '+   <h2>children <b>text</b></h2>\n' +
+        '    <div data-ssr-mismatch-padding-after="1" />\n' +
```

Comment:
Changes to the diff printer compared to PR #12063:
1) always prints elements with no children as self-closing, assuming the diff output is renderer-agnostic and is presented in JSX-like syntax, not in the underlying HTML syntax; previously the diff printer had access to `omittedCloseTags` from `react-dom`, now the diff printer is renderer-agnostic so cannot rely on that.
2) applies some logic to determine if plain strings should be printed in curly braces and quotes, or as plain text (see `printChildrenValue` and `shouldPrintStringAsRawUnescapedText` in `ReactFiberHydrationWarning.js`) to produce more natural, less cluttered output.

No change:
- the HTML comment nodes and other non-hydratable host nodes are still printed out to add more context to the diff (`react-reconciler` asks `react-dom` to print them via a `ReactDOMHostConfig` function); this is debatable, you may want to remove them to make the output JSX-only, and simplify some code.